src/riak_core_info_service_process.erl

```diff
@@ -0,0 +1,168 @@
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-ifdef(TEST).
+-compile(export_all).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%% `gen_server' implementation
+
+%% All arguments must be fully-defined callbacks.
+start_link(Registration, Shutdown, Provider, ResponseHandler)
+  when is_tuple(Registration),
+       is_tuple(Shutdown),
+       is_tuple(Provider),
+       is_tuple(ResponseHandler) ->
+    State = #state{registration = Registration,
+                   shutdown = Shutdown,
+                   service_provider = Provider,
+                   response_handler = ResponseHandler},
+    gen_server:start_link(?MODULE, State, []).
+
+init(#state{registration=Registration}=State) ->
+    %% Registration callback *must* return `ok' or this process will
+    %% shut down
+    handle_consumer_response(apply_callback(Registration, [self()]), registration, State).
+
+handle_info({invoke, ProviderParams, HandlerContext}, State) ->
+    handle_invoke_message(ProviderParams, HandlerContext, State);
+
+handle_info(callback_shutdown, State) ->
+    {stop, normal, State}.
+
+terminate(_Reason, State) ->
+    handle_shutdown_message(State).
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+handle_call(Request, _From, State) ->
+    terminate_not_implemented(Request, State).
+
+handle_cast(Request, State) ->
+    terminate_not_implemented(Request, State).
+
+%% Private
+
+terminate_not_implemented(Request, State) ->
+    lager:warning("Terminating - unknown message received: ~p", [Request]),
+    {stop, not_implemented, State}.
+
+handle_invoke_message(ProviderParams, HandlerContext,
+                      #state{service_provider = Provider,
+                             response_handler = ResponseHandler } = State) ->
+    %% At this layer we don't care much whether a callback succeeded
+    {_, Result} = apply_callback(Provider, ProviderParams),
+
+    Reply = {Result, ProviderParams, HandlerContext},
+    handle_consumer_response(apply_callback(ResponseHandler, [Reply]), response, State).
+
+handle_shutdown_message(#state{shutdown = Shutdown}=State) ->
+    _ = apply_callback(Shutdown, [self()]),
+    {stop, normal, State}.
+
+%% The original response to the callback will be wrapped in an `ok' or
+%% `exit' tuple, primarily so that the service process can terminate
+%% if something goes wrong during registration.
+-spec apply_callback(callback(), list(term())) -> {ok, term()} |
+                                                  {error, term()}.
+apply_callback({Mod, Fun, StaticArgs}, CallSpecificArgs) ->
+    %% Invoke the callback, passing static + call-specific args for this call
+    Args = StaticArgs ++ CallSpecificArgs,
+    _ = case (catch erlang:apply(Mod, Fun, Args)) of
+            {'EXIT', {Reason, _Stack}}=Exit->
+                %% provider called erlang:error(Reason)
+                lager:warning("~p:~p/~B exited (~p)",
```

Comment:
`lager` calls, otoh, can return structured results, so dialyzer will flag them when the option to check returns is on.